### PR TITLE
[#4027] Add a mechanism to specify bind address for yb-ctl (only rf=1)

### PR DESCRIPTION
**Pull request by iSignal**
**Summary**: @schoudhury  requested adding a mechanism to interact with a yb-ctl created cluster from a remote server. This diff adds a listen_ip parameter that can be used to specify the network interface that the clients bind to (redis, ycql, ysqlsh, web server). The parameter can only be specified if rf = 1.

**Test Plan**: Run in different cases with listen_ip unspecified, 0.0.0.0, local ip

```
17:22 $ bin/yb-ctl start --listen_ip=10.150.0.148
Starting cluster with base directory /net/dev-server-sanketh-3/share/yugabyte-data
Waiting for cluster to be ready.
----------------------------------------------------------------------------------------------------
| Node Count: 1 | Replication Factor: 1                                                            |
----------------------------------------------------------------------------------------------------
| JDBC                : jdbc:postgresql://10.150.0.148:5433/postgres                               |
| YSQL Shell          : build/latest/bin/ysqlsh -h 10.150.0.148                                    |
| YCQL Shell          : build/latest/bin/cqlsh 10.150.0.148                                        |
| YEDIS Shell         : build/latest/bin/redis-cli -h 10.150.0.148                                 |
| Web UI              : http://10.150.0.148:7000/                                                  |
| Cluster Data        : /net/dev-server-sanketh-3/share/yugabyte-data                              |
----------------------------------------------------------------------------------------------------



17:22 $ bin/yb-ctl restart --listen_ip=0.0.0.0
Stopping cluster.
Starting cluster.
Waiting for cluster to be ready.
----------------------------------------------------------------------------------------------------
| Node Count: 1 | Replication Factor: 1                                                            |
----------------------------------------------------------------------------------------------------
| JDBC                : jdbc:postgresql://127.0.0.1:5433/postgres                                  |
| YSQL Shell          : build/latest/bin/ysqlsh                                                    |
| YCQL Shell          : build/latest/bin/cqlsh                                                     |
| YEDIS Shell         : build/latest/bin/redis-cli                                                 |
| Web UI              : http://127.0.0.1:7000/                                                     |
| Cluster Data        : /net/dev-server-sanketh-3/share/yugabyte-data                              |
----------------------------------------------------------------------------------------------------




17:23 $ bin/yb-ctl restart
Stopping cluster.
Starting cluster.
Waiting for cluster to be ready.
----------------------------------------------------------------------------------------------------
| Node Count: 1 | Replication Factor: 1                                                            |
----------------------------------------------------------------------------------------------------
| JDBC                : jdbc:postgresql://127.0.0.1:5433/postgres                                  |
| YSQL Shell          : build/latest/bin/ysqlsh                                                    |
| YCQL Shell          : build/latest/bin/cqlsh                                                     |
| YEDIS Shell         : build/latest/bin/redis-cli                                                 |
| Web UI              : http://127.0.0.1:7000/                                                     |
| Cluster Data        : /net/dev-server-sanketh-3/share/yugabyte-data                              |
----------------------------------------------------------------------------------------------------```

### CODE REVIEW

File: bin/yb-ctl [hunk 1]
```diff
@@ -1090,6 +1114,12 @@ class ClusterControl:
                 "--ip_start", default=DEFAULT_IP_START, type=int,
                 help="Start index for IP address")
 
+            subparsers[cmd].add_argument(
+                "--listen_ip",
+                dest='listen_ip_dest',
```
Comment (WesleyW) [inline bin/yb-ctl]: I don't think `yb-ctl` ever uses `dest`

Comment (WesleyW) [inline bin/yb-ctl:1119]: Does `yb-ctl` ever uses these `dest` vars? Don't need change just wondering.

File: bin/yb-ctl [hunk 2]
```diff
@@ -1221,7 +1251,7 @@ class ClusterControl:
         command_list += [
             # Add in all the shared flags
             "--fs_data_dirs \"{}\"".format(",".join(node_base_dirs)),
-            "--webserver_interface {}".format(self.options.get_ip_address(daemon_id)),
+            "--webserver_interface {}".format(self.options.get_client_listen_host_port(daemon_id)),
             "--rpc_bind_addresses {}".format(self.options.get_ip_address(daemon_id)),
```
Comment (WesleyW) [inline bin/yb-ctl:1255]: Does this need to be changed as well?

File: bin/yb-ctl [hunk 3]
```diff
@@ -1281,7 +1313,8 @@ class ClusterControl:
         if self.is_ysql_enabled():
             command_list += [
                 "--enable_ysql=true",
-                "--pgsql_proxy_bind_address=" + get_host_port('ysql')
+                "--pgsql_proxy_bind_address=" +
+                self.options.get_client_listen_host_port(daemon_id, 'ysql')
```
Comment (WesleyW) [inline bin/yb-ctl:1317]: Instead of changing these, change the `get_host_port` function at line 1298

File: bin/yb-ctl [hunk 4]
```diff
@@ -1090,6 +1114,12 @@ class ClusterControl:
                 "--ip_start", default=DEFAULT_IP_START, type=int,
                 help="Start index for IP address")
 
+            subparsers[cmd].add_argument(
+                "--listen_ip",
```
Comment (WesleyW) [inline bin/yb-ctl:1118]: `bind_ip` would probably be better ?

**Review by WesleyW — APPROVED** [review]

File: bin/yb-ctl (hunk 4 above)
Comment (WesleyW) [reply bin/yb-ctl:1118]: Shouldn't this also be included in the `stop` subparsers? Does `yb-ctl stop` work if you have `yb-ctl stop --listen_addr 11.11.11.11`?

File: bin/yb-ctl (hunk 1 above)
Comment (iSignal) [reply bin/yb-ctl:1119]: No it doesn't use the dest, the value gets filled into args.listen_ip

File: bin/yb-ctl (hunk 2 above)
Comment (iSignal) [reply bin/yb-ctl:1255]: I intentionally didn't want to change rpc_bind_address (coz then I would need to supply server_broadcast_addresses in these scenarios)

File: bin/yb-ctl (hunk 4 above)
Comment (iSignal) [reply bin/yb-ctl:1118]: stop works fine without the listen_addr because it searches for procs based on rpc_bind_address and I didn't touch that.

**Comment by iSignal** [comment]
There are intermittent errors with yb-ctl-test.sh. Ran them locally on my host and they pass. Will take a look at them later.

**Pull request merged**